### PR TITLE
added toLWTH conversion if startLocation is rectLTRB

### LINKED ISSUE #12424: vision Rect conversion bug
### Steps to reproduce:
Open Pycharm and focus on the main text field. Open Python Console and do the following:

`from vision.util import getContextRect
from vision.constants import Context
getContextRect(Context.CARET, focus)`

### Actual behavior:
The following error occurs:
AttributeError: 'RectLTRB' object has no attribute 'toLTRB'
### Expected behavior:
return a RectLTRB of the current caret position.
#### NVDA installed/portable/running from source:
running from source, but also occurs on installed and portable versions
#### NVDA version:
2020.4
#### Windows version:
windows 10
#### Name and version of other software in use when reproducing the issue:
NA
#### Does the issue still occur after restarting your computer?
yes
#### Have you tried any other versions of NVDA? If so, please report their behaviors.
no
#### If add-ons are disabled, is your problem still occurring?
yes
#### Does the issue still occur after you run the COM Registration Fixing Tool in NVDA's tools menu?
yes

The bug is kinda obvious when looking at the code. There is just a line `rect.ToLTRB()` and it does not check what kind of rectangle it is. In most cases this rect is a rectLTWH so ToLTRB makes sense.
It is easily fixed by an if statement checking whether the rectangle is already LTRB. However, it might be more robust to add the function ToLTRB() to LocationHelper.rectLTRB that returns itself. I kinda suspect that there are more of these bugs lurking arround in the vision framework and just plugging a function ToLTRB() into rectLTRB would spare some if-statements in the vision framework.

### PR DESCRIPTION
**Pull request by FalkoBabbage**
### Link to issue number:
Fixes https://github.com/nvaccess/nvda/issues/12424
### Summary of the issue:
JABTextInfo can return rectLTRB which is then used in the boudingRects property. this can be an issue for functions that expect rectLTWH types to be returned.
### Description of user facing changes
None
### Description of development approach
walked through where JABTextInfo could return a rectLTRB instance. found it
### Testing strategy:
Ran runnvda.bat with adjusted code. Followed the steps in issue 12424 to reproduce the error and got a valid answer instead.
### Known issues with pull request:
-
### Change log entries:
New features
Changes
Bug fixes
For Developers
-offsets.OffsetsTextInfo._get_boundingRects return type is specified to List[locationHelper.rectLTWH] as conform to textInfos.TextInfo._get_boundingRects. offsets.OffsetsTextInfo._get_boundingRects is adjusted to always return the specified type.
### Code Review Checklist:

- [ ] Pull Request description:
  - description is up to date
  - change log entries
- [ ] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [ ] API is compatible with existing add-ons.
- [ ] Documentation:
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [ ] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [ ] Security precautions taken.

### CODE REVIEW

**Comment by seanbudd** [comment]
Hi @FalkoBabbage, I don't think that change log description is appropriate for bug fixes.
If there are user facing changes, list them there.
Otherwise can we put in "changes for developers"?

I'm also not sure if this description is specific enough:

>-OffsetsTextInfo.boundingRects sometimes returned rectLTRB instances in some JAB textInfo objects, this is no longer the case. 

What does it return instead now?

Can you add typing information to `_get_boundingRects`?

Considering that `rectLTRB` was returned previously, will changing this return type be an API breakage?
What do other parts of NVDA expect as a type here?

**Comment by FalkoBabbage** [comment]
Hi @seanbudd. 
I agree that this should be in "changes for developers", since the bugs that it fixes occur later down the line.

> I'm also not sure if this description is specific enough:

The origin of the error that I occurred is as follows:
1. I try to retrieve the boundingRects property of a  JAB text info. The `_get_boundingRects` method that is used is from textInfos.offsets.OffsetsTextInfo
2. the `_get_boundingRects` method uses `_getBoundingRectFromOffset` for with the startoffset to get the startLocation. If the endoffset is bigger than startOffset + 1, all rectangles that are obtained from `_getBoundingRectFromOffset` for each offset are put through the function locationHelper.RectLTWH.fromCollection, making it automatically a rectLTWH.
3. If the endoffset is not bigger than startOffset + 1, the startLocation is only checked if this is a point. If it is not a point, the start location is put into the rects list as is.
4.  `_getBoundingRectFromOffset` does not have any type hinting. In JABTextInfo in NVDAObjects/JAB/__init__.py this function explicitly returns a rectLTRB. 

Maybe the following is better to describe the changes:
"_get_boundingRects" is assumed to return locationHelper.rectLTWH types, but for textInfos.offsets.OffsetsTextInfo there there can be situations where a instance of locationHelper.rectLTRB is returned. This is solved by adding an explicit conversion to rectLTWH on the location where a rectLTRB can occur.

> Can you add typing information to _get_boundingRects? 

 In textInfos/__init__.py the _get_boundingRects method of the class TextInfo already specifies the return type locationHelper.RectLTWH in its comment. This is not enforced by the typing module and is just a comment.

> Considering that rectLTRB was returned previously, will changing this return type be an API breakage?
> What do other parts of NVDA expect as a type here?

Since `_get_boundingRects` does not have the return type enforced there indeed might be pieces of code that expect a rectLTRB, But considering the indicated return type of `_get_boundingRects` they shouldn't and therefore I wouldn't consider it a breaking API change.

**Comment by seanbudd** [comment]
Thanks for explaining the bug in more detail.

>In textInfos/init.py the _get_boundingRects method of the class TextInfo already specifies the return type locationHelper.RectLTWH in its comment. This is not enforced by the typing module and is just a comment.

Can you please add typing information to the function header here. For example `def func(self) -> List[locationHelper.RectLTWH]:`

https://github.com/nvaccess/nvda/blob/474f22863749748a6f0ccd1bc78c9d4572bd8497/source/textInfos/offsets.py#L182

>Since _get_boundingRects does not have the return type enforced there indeed might be pieces of code that expect a rectLTRB, But considering the indicated return type of _get_boundingRects they shouldn't and therefore I wouldn't consider it a breaking API change.

Ah I see, makes sense. Thanks for explaining

**Comment by FalkoBabbage** [comment]
> Can you please add typing information to the function header here. For example def func(self) -> List[locationHelper.RectLTWH]:

_get_boundingRects is defined in textInfos.TextInfo. I suggest to add the correct typing there if that is desired since I assume that any override would also be "forced" to use this type. Should 
I also add typing for all other overrides for _get_boundingRects?

**Comment by seanbudd** [comment]
I think it would be sufficient to just add it for the function you updated.

**Comment by FalkoBabbage** [comment]
I have looked into reducing the complexity but as far as I know that would mean that I have to split up the function into different subfunctions. In that case we would end up with more methods in the OffsetsTextInfo class which should also make sense for its child-classes downstream.
Besides, I am not sure why appveyor flags my code as to complex. I have already reduced the existing complexity of this function by removing an if/else statement.

**Comment by CyrilleB79** [comment]
> Besides, I am not sure why appveyor flags my code as to complex. I have already reduced the existing complexity of this function by removing an if/else statement.

The linter only checks the code that you have modified. And among others, it checks the complexity of the functions that you have modified. It does not care if you reduced the complexity, it just check if it is too high or not. The original function was very too complex. But it has probably been written before the linter check was introduced.

You have two options:
* Either you split the function to reduce its complexity
* Or (if NVAccess agrees with it), given you have modified only few lines of this function, you may add a `#noqa` directive to ignore the complexity.

If it's not too much work and depending on the impact of the modifications, the first solution is preferred. This allows the code to converge towards a linted code.

**Comment by seanbudd** [comment]
Either option @CyrilleB79 suggested here is appropriate. Search for `C901` for functions handled in a similar way.

File: source/textInfos/offsets.py [hunk 1]
```diff
@@ -179,7 +180,7 @@ def _get_locationText(self):
 			textList.append(_("at {x}, {y}").format(x=curPoint.x,y=curPoint.y))
 		return ", ".join(textList)
 
-	def _get_boundingRects(self):
+	def _get_boundingRects(self) -> List[locationHelper.RectLTWH]:  # noqa: C901
```
Comment (seanbudd) [inline source/textInfos/offsets.py:183]: please see similar comment in the NVDA code base and apply the full message

**Review by seanbudd — APPROVED** [review]
Thanks @FalkoBabbage

**Pull request merged**